core/src/test/scala/unit/kafka/log/ProducerStateManagerTest.scala

```diff
@@ -834,6 +834,40 @@ class ProducerStateManagerTest {
     assertEquals(None, stateManager.lastEntry(producerId).get.currentTxnFirstOffset)
   }
 
+  @Test
+  def testRemoveStraySnapshotsKeepCleanShutdownSnapshot(): Unit = {
+    // Test that when stray snapshots are removed, the largest stray snapshot is kept around. This covers the case where
+    // the broker shutdown cleanly and emitted a snapshot file larger than the base offset of the active segment.
+
+    // Create 3 snapshot files at different offsets.
+    Log.producerSnapshotFile(logDir, 42).createNewFile()
+    Log.producerSnapshotFile(logDir, 5).createNewFile()
+    Log.producerSnapshotFile(logDir, 2).createNewFile()
+
+    // claim that we only have one segment with a base offset of 5
+    stateManager.removeStraySnapshots(Set(5))
+
+    // The snapshot file at offset 2 should be considered a stray, but the snapshot at 42 should be kept
+    // around because it is the largest snapshot.
+    assertEquals(Some(42), stateManager.latestSnapshotOffset)
+    assertEquals(Some(5), stateManager.oldestSnapshotOffset)
+    assertEquals(Seq(5, 42), ProducerStateManager.listSnapshotFiles(logDir).map(_.offset).sorted)
+  }
+
+  @Test
+  def testRemoveAllStraySnapshots(): Unit = {
+    // Test that when stray snapshots are removed, all stray snapshots are removed when the base offset of the largest
+    // segment exceeds the offset of the largest stray snapshot.
```

Comment:
Below, the base offset of the largest segment equals to and doesn't exceed the offset of the largest stray snapshot.

Reply:
I think this sentence is a bit confusing. Snapshot 42 is not meant to be a stray snapshot here, only 5 and 2 are. I will try to reword this.

Reply:
Hmm, I think my comment here could be worded better. Offset `42` here is not a "stray", since we provide it along with the list of segmentBaseOffsets to `removeStraySnapshots`. 

I'll change up the wording on this, thanks!